File: LeetCode/Linked Lists/23-merge-k-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function(lists) {
    let length = lists.length;
    
    let interval = 1;
    while (interval < length) {
        for (let i = 0; i < length; i += interval * 2) {
            lists[i] = mergeTwoLists(lists[i], lists[i+interval])
        }
        interval *= 2;
    }
    
    return length > 0 ? lists[0] : null;
    
};

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var mergeTwoLists = function(l1, l2) {
    let list = new ListNode(" ");
    let ptr = list;
    while (l1 && l2) {
        if (l1.val < l2.val) {
            list.next = l1;
            l1 = l1.next;
            list = list.next;
        } else {
            list.next = l2;
            l2 = l2.next;
            list = list.next;
        }
    }
    if (l1) {
        list.next = l1;
    } else {
        list.next = l2
    }
    return ptr.next;
};

File: LeetCode/Sorting/23-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        if not lists:
            return None

        def merge(list1, list2):
            if not list2:
                return list1
            head = ListNode(-1)
            node = head
            while list1 and list2:
                if list1.val < list2.val:
                    node.next = list1
                    list1 = list1.next
                else:
                    node.next = ListNode(list2.val)
                    list2 = list2.next
                node = node.next
            if list1:
                node.next = list1
            else:
                node.next = list2
            return head.next

        j = len(lists)
        while j > 1:
            i = 0
            while 2 * i < j:
                lists[i] = merge(lists[2*i], lists[2*i+1]
                                 if 2*i+1 < j else None)
                i += 1
            j = (j+1) // 2

        return lists[0]
